openslides/utils/consumers.py

```diff
@@ -0,0 +1,306 @@
+from typing import Any, Dict, List, Optional
+
+from asgiref.sync import sync_to_async
+from channels.db import database_sync_to_async
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+
+from ..core.config import config
+from ..core.models import Projector
+from .auth import a_anonymous_is_enabled, has_perm
+from .cache import element_cache, split_element_id
+from .collection import AutoupdateFormat  # noqa
+from .collection import (
+    Collection,
+    CollectionElement,
+    format_for_autoupdate,
+    from_channel_message,
+)
+
+
+class SiteConsumer(AsyncJsonWebsocketConsumer):
+    """
+    Websocket Consumer for the site.
+    """
+    groups = ['site']
+
+    async def connect(self) -> None:
+        """
+        A user connects to the site.
+
+        If it is an anonymous user and anonymous is disabled, the connection is closed.
+
+        Sends the startup data to the user.
+        """
+        # TODO: add a way to ask for the data since a change_id and send only data that is newer
+        if not await a_anonymous_is_enabled() and self.scope['user'].id is None:
+            await self.close()
+        else:
+            await self.accept()
+            data = await startup_data(self.scope['user'])
+            await self.send_json(data)
+
+    async def receive_json(self, content: Any) -> None:
+        """
+        If we recieve something from the client we currently just interpret this
+        as a notify message.
+
+        The server adds the sender's user id (0 for anonymous) and reply
+        channel name so that a receiver client may reply to the sender or to all
+        sender's instances.
+        """
+        if notify_message_is_valid(content):
+            await self.channel_layer.group_send(
```

Comment:
Can you send in parallel?

Reply:
Maybe. But this is not important, because I want to delete the projector consumer anyway.